src/conv-core/convcore.C

```diff
@@ -2888,27 +2893,65 @@ void CmiGroupReduceStruct(CmiGroup grp, void *data, CmiReducePupFn pupFn,
 }
 
 void CmiNodeReduce(void *msg, int size, CmiReduceMergeFn mergeFn) {
-  CmiReduction *red = CmiGetNextNodeReduction(CmiNumNodeSpanTreeChildren(CmiMyNode()));
+  const CmiReductionID id = CmiGetNextNodeReductionID();
```

Comment:
Does the call to `CmiGetNextNodeReductionID()` need to be inside the lock, too? Won't the reads and writes to `_nodereduce_seqID_global` race? I see that `id` is used to get the right lock, so perhaps I'm missing something here.

Reply:
`CmiGetNextNodeReductionID()` is only called from `CmiNodeReduce()` and `CmiNodeReduceStruct()`, which are documented as only valid to call on rank 0. I've added assertions to that effect.

Reply:
That ended up being incorrect. These may be called from any rank so long as the node only calls them once. I've made `_nodereduce_seqID_global` `std::atomic` to compensate.